<!DOCTYPE html>
<html lang="en-GB">

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  
  
  <title>Lambda reflections</title>
  <meta name="description" content="Functional Programming, Proofs, Bioinformatics and more.">
  

  <link rel="stylesheet" href="/assets/main.css">
  <link rel="canonical" href="https://nlightnfotis.github.io/posts/2/">
  
  
  <link rel="alternate" type="application/rss+xml" title="Lambda reflections" href="https://nlightnfotis.github.io/feed.xml">

  

  
  <meta property="og:title" content="Lambda reflections">
  <meta property="og:site_name" content="Lambda reflections">
  <meta property="og:url" content="https://nlightnfotis.github.io/posts/2/">
  <meta property="og:description" content="Functional Programming, Proofs, Bioinformatics and more.">
  
  
  <meta name="twitter:card" content="summary">
  <meta name="twitter:site" content="NlightNFotis">
  <meta name="twitter:title" content="Lambda reflections">
  <meta name="twitter:description" content="Functional Programming, Proofs, Bioinformatics and more.">
  
    <meta name="twitter:creator" content="NlightNFotis">
  
  

  <link rel="dns-prefetch" href="https://fonts.gstatic.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Bitter:ital,wght@0,400;0,700;1,400&amp;display=swap" rel="stylesheet">

  

</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/">Lambda reflections</a>

    <nav class="site-nav">
      
        
        <a class="page-link" href="/about/">About</a>
      
        
        <a class="page-link" href="/archives/">Archives</a>
      
        
        <a class="page-link" href="https://github.com/NlightNFotis">GitHub</a>
      
    </nav>

  </div>

</header>


    <main class="page-content" aria-label="Content">
      <div class="wrapper">
        <div class="home">

  

  

  <ul class="post-list">
    
      
      

      <li>
        <header class="post-header">
          <h1 class="post-title">
            
              <a class="post-link" href="/2022/05/12/reflections-on-learning-how-to-grow-almost-anything/">Reflections on Learning &quot;How to Grow (Almost) Anything&quot;</a>
            
          </h1>

          <p class="post-meta">
            May 12, 2022
            
              

 ‚Ä¢
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/synbio/">synbio</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/course/">course</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/mooc/">mooc</a>
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  


            
            
          </p>
        </header>

        <div class="post-content">
          <p>Wow, what a journey this has been!</p>

<p>Yesterday, Wednesday 11th of May, was the final project presentations for committed
listeners in MIT‚Äôs synthetic biology course,
<a href="https://htgaa2022.notion.site/htgaa2022/HTGAA-2022-d39e5560ad83483ab87d415f085b60c6">How To Grow (Almost) Anything</a>,
class of 2022.</p>

<p>When I started the course, sometime in the end of January, I had no idea what a blast
this would be (nor did I have any idea of how much pressure this would apply on me, but
hey, I survived this, and then some üòÅ ).</p>

<p>I credit this class with taking me from rudimentary biology knowledge to a fairly solid
understanding of molecular biology and immunology, all in about 3-4 months.</p>

<p>As you can probably imagine, this was‚Ä¶ intense.</p>

<p>But before I dive more into the experience and what I learnt in this course, let me
give some background into the course and the context of my participation.</p>

<h2>Some background</h2>

<p>Anyone who knows me is aware of my awe at the potential of synthetic biology. There‚Äôs
no doubt in my mind that it‚Äôs going to be the next exponential growth sector after
the computer industry. The field itself is fairly nascent, but it‚Äôs going to pick
up the pace in the next 5-10 years. And although I believe that all its subdomains
hold massive potential, I‚Äôm especially keen on synthetic biology applications in
human therapeutics.</p>

<p>Now, I keep it no secret that I want to found a synthetic biology company in the
(hopefully, not so distant) future.</p>

<p>Given that, you may guess that in general I was observant of any opportunities
to dig a bit deeper into the field.</p>

<p>With that in mind (along with a hefty bit of luck - fate? üòÅ ), I came across a
tweet by the organiser of the course, <a href="https://twitter.com/davidsunkong">David Kong</a>,
advertising the course, along with the fact that this year it would admit a small
cohort of students who want to attend remotely.</p>

<p>Wasting no time, I shot an email towards David, and was surprised when I got an
email a couple days later admitting me to the course, and sharing zoom links for
the course‚Äôs first lecture.</p>

<h2>The Attendance Experience</h2>

<p>The course was setup in a set of two classes per week:</p>

<ul>
  <li>one three-hour lecture, in which the first hour was spent reviewing homework
from last week, and the next two hours had different presenters give presentations
on various topics in synthetic biology, and</li>
  <li>a two-hour class the next day, that included a recitation session - which was
previewing next week‚Äôs homework - along with a homework review for online listeners.</li>
</ul>

<p>To my surprise, there was a relatively high number of online students participating
(~ 45, IIRC - about 3 times the number of MIT students physically attending), but
only about a third of the online students made it through to the end.</p>

<p>Attendance was taken every lecture, for both online listeners and the students
physically present at MIT. We were also expected to do all the exercises for
the week - with the TAs ensuring that this was the case.</p>

<p>In the middle of the course‚Äôs duration, we also had a handful of hands-on workshops
(of which we had to attend some), aptly named the <em>Festival of Growing</em>.</p>

<h2>The Lectures</h2>

<p>The course was setup around a number of themes (therapeutics, biofabrication/biomaterials,
DNA/RNA editing, microfluidics, etc), and every week we had two (sometimes three)
different presenters (which ranged from university professors to company CEOs) present
on the topic at hand.</p>

<p>The first couple of weeks were focused on teaching the foundations of molecular biology
(but in a very, <em>very</em> accelerated timeframe), and then the following weeks assumed
competence in molecular biology and built on top of that in terms of presenting the
state of the art in applications in various domains.</p>

<p>It‚Äôs probably worth noting that the course attracted people from a wide variety
of backgrounds - you could find architects, designers, computer scientists, chemists,
etc. As you can imagine, a lot of these people didn‚Äôt have a background in biology
(or had a very shallow one, myself included), and the intensity of the ramp up of
the first few weeks might have contributed to the dropouts.</p>

<p>For the MIT/Harvard students, most of these lectures were also followed by a lab
component. For online listeners, we would partially follow along those processes
(given the lack of wet lab access for some of us), or, at other times, we would
be given substitute exercises that involved using online tools (such as <a href="https://www.benchling.com/">Benchling</a>
to simulate the in-vitro experiments in-silico.</p>

<h2>The <em>Festival of Growing</em></h2>

<p>Sometime around the half of the semester, we had a pause in the online lectures
for a couple of weeks in favour of attending various workshops. Those workshops
were structured around a number of different themes. Examples of those were growing
mycelium at home, or performing DNA sequencing using a nanopore device.</p>

<p>From these workshops, some of them were on-premise and some of them were online.
We did manage to watch one of the on-premise workshops (the nanopore sequencing one)
via a TA streaming the workshop, which was nice. In retrospect, this would be
a good idea for all of the on-premise workshops - maybe that will happen from
next year onwards?</p>

<p>For both online and MIT/Harvard students, we had to attend at least 1 of the workshops
(attendance was taken), with no upper limit on how many you could attend.</p>

<p>I think the Festival of Growing was a nice little fragment of the course‚Äôs structure,
and it also contributed in making the course feel a lot more dynamic in nature
(compared to an alternative, hypothetical course consisting only of lectures).</p>

<h2>The Final Project</h2>

<p>A relatively big part of the course (and the main focus of the last two weeks)
was the the preparation of the final project. It had been made clear from the
start that this was an integral part of the course, and that everyone who stuck
to the end (either online listener or on-premise student) would have to
present one.</p>

<p>We had a lot of leeway in terms of the project that we would choose to do -
though you were <em>strongly encouraged</em> to pick a project topic that fell within
the experience/knowledge range of available TAs. For the on-premise students,
the idea is that the project would include a lab component but for online students
this was considered optional (given that only some people had access to a lab),
but we still had to present on a topic that was considered feasible (if speculative),
and the expectation was that the project would be fairly well-researched.</p>

<p>For each one of us doing a final project, we would be assigned a mentor (a
TA or industry professional) who would end up as an endpoint for answering
questions/providing guidance on our project.</p>

<h2>My Own Experience</h2>

<p>I have to start with the fact that I enjoyed all of the course, and that
I‚Äôm grateful to MIT/the organisers for admitting us online listeners to the
course.</p>

<p>Having said that, the first couple of weeks were very hard - it required
<strong>a lot</strong> of outside studying (we‚Äôre talking several textbooks worth of
reading) to catch up to the required knowledge. If you have to juggle a
job as well, and you‚Äôre not a full-time student, that can definitely feel
pretty overwhelming at times.</p>

<p>(By the way, this is coming from someone who already commits 3-4 hours of
study on a daily basis, if that helps underline the above paragraph - though
to be fair, I think that a lot of the pressure was coming from the tight
weekly deadlines for exercises.)</p>

<p>I recall that at one point during the first two weeks I felt the presenters
were speaking something other than English. üòì The nice thing is that persistence
does pay of, however, and by the end, you are also capable of fluently conversing
with everyone else in this weird little dialect of English.</p>

<p>The festival of growing was also very nice, but, personally, I was a little bit
miffed by the fact that the workshop I was most looking forward to (metabolic
simulation in Python) got cancelled without any explanation a day before it
was to be presented.</p>

<p>For me the most enjoyable part of the course was the final project because
by that point I already knew enough to be able to do some competent background
research on the project.</p>

<h2>In Conclusion</h2>

<p>In this final bit, I want to say that I wholeheartedly recommend this course
to anyone who is keen on synthetic biology, if only to see some of the people
on the forefront of the domain discuss the possibilities and what they are
currently exploring.</p>

<p>I also have a feeling that David and his course are going to be a future kingmaker
of synthetic biology üòâ You would hear very often during the presentations (person
X, CEO of Y, and alumni of HTGAA few years ago). I think this trend is going to
continue for many more years to come.</p>

<p>Again many thanks to the organisers (<strong>David</strong> in particular), the lecturers, TAs,
and everyone else who contributed to this course.</p>

<p>Special thanks from my end to the following people as well:</p>

<ul>
  <li><strong>Jessica Weber</strong>, for entertaining a barrage of questions I posed during
one of the lectures.</li>
  <li><strong>Karolina Sulich</strong>, for providing some early assistance during my project,
and for introducing me to Amy Holt,</li>
  <li><strong>Amy Holt</strong>, who provided me with specific answers to some immunology related
questions I had, with regard to my final project.</li>
</ul>

<p>You can find my course notes, exercises, and final project at my course notes
<a href="https://www.notion.so/HTGAA-2022-a43e0f0b56a444f29aea3a94a388c7bf">Notion page</a>.</p>

        </div>
        
      </li>
    
      
      

      <li>
        <header class="post-header">
          <h1 class="post-title">
            
              <a class="post-link" href="/2022/03/28/bioinformatics-stronghold-nucleotide-counting/">Nucleotide Counting the TDD Way - A Bioinformatics Stronghold Story</a>
            
          </h1>

          <p class="post-meta">
            Mar 28, 2022
            
              

 ‚Ä¢
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/ruby/">ruby</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/bioinformatics/">bioinformatics</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/tdd/">tdd</a>
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  


            
            
          </p>
        </header>

        <div class="post-content">
          <p>A few days ago, I was going through my code archives and came across my old solutions
of the <a href="https://rosalind.info/problems/list-view/">Bioinformatics Stronghold</a> by
<a href="https://rosalind.info/problems/locations/">Project Rosalind</a>. For those of you who
don‚Äôt know, <em>Project Rosalind</em> is one of those problem-based skill development exercises,
akin to <a href="https://projecteuler.net">Project Euler</a> but aimed at nascent bioinformaticians.</p>

<p>Coincidentally, I finished a <a href="https://www.goodreads.com/book/show/33527186-test-driven-development-in-ruby">book on Test-Driven Development (TDD) in Ruby</a>
around the same time. I particularly enjoyed TDD the way it was presented in
the book, and I was looking for some problems to apply it to and the Stronghold
Project is the perfect lab space for me to apply those new ideas (because of
its relatively simple from an algorithmic perspective problems).</p>

<hr />

<p>My initial foray into the Stronghold project was in an attempt at comparative solutions,
wherein I attempted the exercises in a number of different programming languages
(Python, Go, OCaml, Racket, etc.) while observing any differences in the style I
chose, refactoring them, benchmarking them, and all around having some good fun
with those üòÑ</p>

<p>One thing I‚Äôm embarrassed to admit about that first attempt though, is that while
my solutions worked (or so I can conveniently recall), they didn‚Äôt contain any
reproducible documentation of their satisfying of the requirements of the
exercise - there were no assertions, no tests, nothing. üòÖ I can only attribute it
to my enthusiasm in getting each solution done to move on to the next one.</p>

<p>Bad me.</p>

<p>I‚Äôm now revisiting these exercises to atone for my insolence. I‚Äôm going to go through
the exercises again, but this time I‚Äôm going to be approaching them in a TDD/BDD style.</p>

<p>But before I move on to actual code, what on Earth is TDD (and its cousin, BDD)? We
already saw that <em>TDD</em> stands for <em>Test-Driven Development</em>. By that, we mean a
programming approach that encourages writing tests for the feature code <strong>before</strong>
the actual feature code is written, so that the tests guide the design of the code
itself.</p>

<p><em>Behaviour-Driven Development</em> (BDD for short), is a sister approach to TDD, but for
the purposes of this article, BDD is the approach of <em>writing the tests in a way
that they reflect prose specification of the behaviour of the module</em> under test.</p>

<p>Okay, are you ready now? Let‚Äôs roll.</p>

<hr />

<p>I‚Äôm going to start with the <a href="https://rosalind.info/problems/dna/">first exercise</a>,
which is about counting nucleotides in a DNA strand, which is given as an input of
a <code class="language-plaintext highlighter-rouge">string</code> type.</p>

<p>Now, if you visit the problem page, you will see that it also gives us a sample
dataset, along with the expected output for that.</p>

<p>(You may have noticed that the above sentence is just a particularly verbose way
of describing a <em>test case</em>. üòâ )</p>

<p>Our implementation language for this exercise is going to be Ruby, for two reasons.</p>

<p>1) I started learning Ruby recently and I‚Äôve been enjoying it a lot, and
2) Ruby comes with excellent built-in support for testing in the form of the <code class="language-plaintext highlighter-rouge">Minitest</code>
   library.</p>

<p>Let‚Äôs quickly make a file called <code class="language-plaintext highlighter-rouge">counting_nucleotides.rb</code>, and add an empty test
specification:</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">require</span> <span class="s1">'minitest/autorun'</span>

<span class="n">describe</span> <span class="s1">'nucleotide counting function'</span> <span class="k">do</span>
<span class="k">end</span>
</code></pre></div></div>

<p>At this point it‚Äôs worth having a pause to think about our test cases before we move
forward.</p>

<p>We already have been given a sample input, as we mentioned above, that we could use
as our test case. The problem, however, with that particular input is that it describes
the functionality of the module when it‚Äôs finished.</p>

<p>That‚Äôs probably a bit <em>too</em> elaborate for us to use now that we start designing our
function.</p>

<p>We probably want something much simpler - indeed, this is what TDD as an approach
is advocating. What might be simpler for us?</p>

<p>What about a strand with a very small length? Say, 8 bases long?</p>

<p>Sounds like it <em>should</em> work.</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">require</span> <span class="s1">'minitest/autorun'</span>

<span class="n">describe</span> <span class="s1">'nucleotide counting function'</span> <span class="k">do</span>
  <span class="n">it</span> <span class="s1">'returns a count of 2 2 2 2 for strand "ATCGATCG"'</span> <span class="k">do</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s1">'ATCGATCG'</span>
    <span class="n">nucleotide_count</span> <span class="o">=</span> <span class="s1">'2 2 2 2'</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>

    <span class="n">assert_equal</span> <span class="n">nucleotide_count</span><span class="p">,</span> <span class="n">result</span>
  <span class="k">end</span>
<span class="k">end</span>
</code></pre></div></div>

<p>Looks good for a first test. Let‚Äôs run it and see what happens.</p>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 64068

<span class="c"># Running:</span>

E

Finished <span class="k">in </span>0.000265s, 3773.5860 runs/s, 0.0000 assertions/s.

  1<span class="o">)</span> Error:
nucleotide counting <span class="k">function</span><span class="c">#test_0001_returns a count of 2 2 2 2 for strand "ATCGATCG":</span>
NoMethodError: undefined method <span class="sb">`</span>count_nucleotides<span class="s1">' for #&lt;#&lt;Class:0x000000015e831da0&gt;:0x000000015b940028&gt;
    counting_nucleotides.rb:8:in `block (2 levels) in &lt;main&gt;'</span>

1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
</code></pre></div></div>

<p>Ahh, it complains that we ‚Äúforgot‚Äù to define our function <code class="language-plaintext highlighter-rouge">count_nucleotides</code>.
Easy to fix, let‚Äôs add a function called that, taking in one parameter, but
with an empty body.</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">require</span> <span class="s1">'minitest/autorun'</span>

<span class="k">def</span> <span class="nf">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>
<span class="k">end</span>

<span class="n">describe</span> <span class="s1">'nucleotide counting function'</span> <span class="k">do</span>
  <span class="n">it</span> <span class="s1">'returns a count of 2 2 2 2 for strand "ATCGATCG"'</span> <span class="k">do</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s1">'ATCGATCG'</span>
    <span class="n">nucleotide_count</span> <span class="o">=</span> <span class="s1">'2 2 2 2'</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>

    <span class="n">assert_equal</span> <span class="n">nucleotide_count</span><span class="p">,</span> <span class="n">result</span>
  <span class="k">end</span>
<span class="k">end</span>
</code></pre></div></div>

<p>Nice, let‚Äôs run it again, and see what we get.</p>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 20901

<span class="c"># Running:</span>

F

Finished <span class="k">in </span>0.000287s, 3484.3209 runs/s, 3484.3209 assertions/s.

  1<span class="o">)</span> Failure:
nucleotide counting <span class="k">function</span><span class="c">#test_0001_returns a count of 2 2 2 2 for strand "ATCGATCG" [counting_nucleotides.rb:12]:</span>
Expected: <span class="s2">"2 2 2 2"</span>
  Actual: nil

1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
</code></pre></div></div>

<p>Okay, this seems a bit more intriguing. Now it doesn‚Äôt come back to us with an <strong>error</strong>.
Rather, it comes back with a <strong>failure</strong>, indicating that the test got executed,
but the expected and the actual results differ. In our case, we asserted in our test
that we expect the result to be a string with the contents <code class="language-plaintext highlighter-rouge">"2 2 2 2"</code> in exactly
that form (spaces and everything), but we got back <code class="language-plaintext highlighter-rouge">nil</code> from the actual execution.</p>

<p>The reason for the <code class="language-plaintext highlighter-rouge">nil</code> in particular is that ruby is an <em>expression-based language</em>.</p>

<p>In an expression-based language every program fragment is an <em>expression</em>, meaning
that it will return a value upon execution of that program fragment (we call that,
in more technical terms, <em>expression evaluation</em>).</p>

<p>A function, thus, is also an expression, and will upon evaluation return the value
of the last expression in its body. For a function with an empty body, there are no
such expressions, so a default of <code class="language-plaintext highlighter-rouge">nil</code> is returned.</p>

<hr />

<p>Right, so we run our test, and got back <code class="language-plaintext highlighter-rouge">nil</code> for a return value. This is our clue
that our test is executing the function as we expect, but our function is not yet
implemented (does not contain a function body). Let‚Äôs crack on with that.</p>

<p>In Ruby, there‚Äôs a very convenient method defined on the <code class="language-plaintext highlighter-rouge">String</code> type whose job
is to count the presence of particular subsequences (substrings in our case).</p>

<p>To no one‚Äôs suprise, it‚Äôs called <code class="language-plaintext highlighter-rouge">count</code>.</p>

<p>Let‚Äôs use that to count the number of nucleotides and return that in a string format.</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>
    <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'A'</span><span class="p">)</span> <span class="o">+</span> <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'C'</span><span class="p">)</span> <span class="o">+</span> <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'G'</span><span class="p">)</span> <span class="o">+</span> <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'T'</span><span class="p">)</span>
<span class="k">end</span>
</code></pre></div></div>

<p>Let‚Äôs run our test again.</p>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 3996

<span class="c"># Running:</span>

F

Finished <span class="k">in </span>0.000411s, 2433.0901 runs/s, 2433.0901 assertions/s.

  1<span class="o">)</span> Failure:
nucleotide counting <span class="k">function</span><span class="c">#test_0001_returns a count of 2 2 2 2 for strand "ATCGATCG" [counting_nucleotides.rb:13]:</span>
Expected: <span class="s2">"2 2 2 2"</span>
  Actual: 8

1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
</code></pre></div></div>

<p>Whoops!</p>

<p>That looks like it found the number of substrings correctly. However, because
of a programming mistake, it looks like it added all the occurences together,
instead of presenting them in a formatted string.</p>

<p>Let‚Äôs do that in the easiest way that comes to mind - using string concatenation
and casting the <code class="language-plaintext highlighter-rouge">integer</code> representing the count back to a <code class="language-plaintext highlighter-rouge">string</code>, while also
adding some spaces (so that we get closer to the expected output):</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>
    <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'A'</span><span class="p">).</span><span class="nf">to_s</span> <span class="o">+</span> <span class="s2">" "</span> <span class="o">+</span> <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'C'</span><span class="p">).</span><span class="nf">to_s</span> <span class="o">+</span> <span class="s2">" "</span> <span class="o">+</span> <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'G'</span><span class="p">).</span><span class="nf">to_s</span> <span class="o">+</span> <span class="s2">" "</span> <span class="o">+</span> <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'T'</span><span class="p">).</span><span class="nf">to_s</span>
<span class="k">end</span>
</code></pre></div></div>

<p>Let‚Äôs see what happens now‚Ä¶</p>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 37259

<span class="c"># Running:</span>

<span class="nb">.</span>

Finished <span class="k">in </span>0.000321s, 3115.2648 runs/s, 3115.2648 assertions/s.

1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
</code></pre></div></div>

<p>It worked! I mean, our code <del>is a bit atrocious</del> could be better, but here
we have our first version of it working. üéâ</p>

<hr />

<p>Now that we have our first unit test passing, let‚Äôs think a bit more about
the test cases we want.</p>

<p>We want at least a representative sample of each of the following:</p>

<ul>
  <li>Positive cases,</li>
  <li>Negative cases,</li>
  <li>Degenerate cases.</li>
</ul>

<p><em>Positive cases</em> are cases in which we exercise the <em>happy path</em> - the code path we
were most anticipating when we were designing our code.</p>

<p><em>Negative cases</em> are cases in which we divert away from the happy path, and
try to exercise error conditions, etc.</p>

<p><em>Degenerate cases</em> are cases in which we test around boundary conditions, such as
empty lists, strings, etc, to see if our function can handle these cases.</p>

<p>We already have a test for a positive case, so right now, it might make more sense
for us to test against a negative case. So what would be a negative case for us?</p>

<p>We are being passed a strand in as a <code class="language-plaintext highlighter-rouge">string</code>. Given that a <code class="language-plaintext highlighter-rouge">string</code> can have a
lot more characters than just the four representing nucleobases, what happens if we
have a string that has characters that don‚Äôt represent a nucleotide base? What happens
if we have something like <code class="language-plaintext highlighter-rouge">ATCGW</code> for a <code class="language-plaintext highlighter-rouge">strand</code>? Let‚Äôs see.</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">it</span> <span class="s1">'throws an exception if a non-base encoding character is found in the strand'</span> <span class="k">do</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s1">'ATCGW'</span>
    
    <span class="n">assert_raises</span><span class="p">(</span><span class="no">ArgumentError</span><span class="p">)</span> <span class="p">{</span> <span class="n">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span> <span class="p">}</span>
  <span class="k">end</span>
</code></pre></div></div>

<p>Let‚Äôs run it to see what happened this time.</p>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 54524

<span class="c"># Running:</span>

.F

Finished <span class="k">in </span>0.000506s, 3952.5692 runs/s, 3952.5692 assertions/s.

  1<span class="o">)</span> Failure:
nucleotide counting <span class="k">function</span><span class="c">#test_0002_throws an exception if a non-base encoding character is found in the strand [counting_nucleotides.rb:19]:</span>
ArgumentError expected but nothing was raised.

2 runs, 2 assertions, 1 failures, 0 errors, 0 skips
</code></pre></div></div>

<p>üò±</p>

<p>Yikes! We were expecting an exception to be raised, but none was raised.
That means that our code had no issue handling the invalid string. Let‚Äôs fix that.</p>

<p>Let‚Äôs fix that the simplest way we can: by defining a list of illegal characters
for the strand string and seeing if they are present in the string.</p>

<p>That gets us with the following version of <code class="language-plaintext highlighter-rouge">count_nucleotides</code>:</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>
    <span class="n">illegal_chars</span> <span class="o">=</span> <span class="s1">'BDEFHIJKLNOPQRSUVWXYZ'</span>
    <span class="n">illegal_chars</span><span class="p">.</span><span class="nf">split</span><span class="p">(</span><span class="s1">''</span><span class="p">).</span><span class="nf">each</span> <span class="k">do</span> <span class="o">|</span><span class="n">char</span><span class="o">|</span>
        <span class="k">if</span> <span class="n">strand</span><span class="p">.</span><span class="nf">include?</span><span class="p">(</span><span class="n">char</span><span class="p">)</span> <span class="k">then</span>
            <span class="k">raise</span> <span class="no">ArgumentError</span><span class="p">.</span><span class="nf">new</span><span class="p">(</span><span class="s1">'Illegal character in strand '</span> <span class="o">+</span> <span class="n">char</span><span class="p">)</span>
        <span class="k">end</span>
    <span class="k">end</span>
    <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'A'</span><span class="p">).</span><span class="nf">to_s</span> <span class="o">+</span> <span class="s2">" "</span> <span class="o">+</span> <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'C'</span><span class="p">).</span><span class="nf">to_s</span> <span class="o">+</span> <span class="s2">" "</span> <span class="o">+</span> <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'G'</span><span class="p">).</span><span class="nf">to_s</span> <span class="o">+</span> <span class="s2">" "</span> <span class="o">+</span> <span class="n">strand</span><span class="p">.</span><span class="nf">count</span><span class="p">(</span><span class="s1">'T'</span><span class="p">).</span><span class="nf">to_s</span>
<span class="k">end</span>
</code></pre></div></div>

<p>Let‚Äôs see where we stand now:</p>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 25460

<span class="c"># Running:</span>

..

Finished <span class="k">in </span>0.000348s, 5747.1265 runs/s, 5747.1265 assertions/s.

2 runs, 2 assertions, 0 failures, 0 errors, 0 skips
</code></pre></div></div>

<p>Nice. That works, so now both our positive case and our negative test cases pass.</p>

<p>The only downside is that we are left with a <code class="language-plaintext highlighter-rouge">counting_nucleotides</code> function that
looks a bit hard to read - not to mention a bit wasteful, too.</p>

<p>(Aside: It loops through the string a lot more times than it needs to, as it loops
once per every illegal character it‚Äôs looking for, and then once for every character
it‚Äôs searching the count for.)</p>

<hr />

<p>At this point, it‚Äôs worth to pause, and reflect on where we are in the process so far.</p>

<p>TDD is a loop of the following 3 steps:</p>

<ol>
  <li>Write a <em>failing test</em>.</li>
  <li>Make the test <em>pass</em>.</li>
  <li><em>Refactor</em> the implementation.</li>
</ol>

<p>(Refactoring is the reorganisation of the code with the aim of improving it with
regard to some metric, say, robustness, readability, performance, etc)</p>

<p>Up until this point, we have been focusing on the first two steps, but did none
of the third one.</p>

<p>We usually refactor once we get some of our implementation done, and <strong>all our
tests are passing</strong>.</p>

<p>In other words, now is as good time as any to refactor our code.</p>

<hr />

<p>Let‚Äôs have a look at our feature code, the <code class="language-plaintext highlighter-rouge">count_nucleotides</code> function.</p>

<p>What if, instead of looping so many times, we looped just once, and collected
both counts and watched out for any illegal character at the same time?</p>

<p>That does sound like it should improve our performance, now, doesn‚Äôt it?</p>

<p>Let‚Äôs go ahead and do this, and see what happens.</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>
    <span class="n">count_a</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_t</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_c</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_g</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="n">strand</span><span class="p">.</span><span class="nf">split</span><span class="p">(</span><span class="s1">''</span><span class="p">).</span><span class="nf">each</span> <span class="k">do</span> <span class="o">|</span><span class="n">base</span><span class="o">|</span>
        <span class="k">if</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'A'</span> <span class="k">then</span>
            <span class="n">count_a</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elsif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'T'</span> <span class="k">then</span>
            <span class="n">count_t</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elsif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'C'</span> <span class="k">then</span>
            <span class="n">count_c</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elsif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'G'</span> <span class="k">then</span>
            <span class="n">count_g</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span>
            <span class="k">raise</span> <span class="no">ArgumentError</span><span class="p">.</span><span class="nf">new</span><span class="p">(</span><span class="s1">'Invalid character in strand '</span> <span class="o">+</span> <span class="n">base</span><span class="p">)</span>
        <span class="k">end</span>
    <span class="k">end</span>

    <span class="s2">"</span><span class="si">#{</span><span class="n">count_a</span><span class="si">}</span><span class="s2"> </span><span class="si">#{</span><span class="n">count_t</span><span class="si">}</span><span class="s2"> </span><span class="si">#{</span><span class="n">count_c</span><span class="si">}</span><span class="s2"> </span><span class="si">#{</span><span class="n">count_g</span><span class="si">}</span><span class="s2">"</span>
<span class="k">end</span>
</code></pre></div></div>

<p>Looks simpler to me. Does it work, though?</p>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 48449

<span class="c"># Running:</span>

..

Finished <span class="k">in </span>0.000378s, 5291.0053 runs/s, 5291.0053 assertions/s.

2 runs, 2 assertions, 0 failures, 0 errors, 0 skips
</code></pre></div></div>

<p>It does.</p>

<p>And just like this, we saw the massive benefit of having automated tests for this:
we did some pretty significant <em>structural changes</em> to our function under test,
but even so, we are confident that its observable behaviour remains unchanged
given our tests and their coverage.</p>

<hr />

<p>Now that we have both the positive and negative tests, can we write a test for
the degenerate case?</p>

<p>Turns out we can.</p>

<p>A degenerate case for us would be an <strong>empty string</strong> (<code class="language-plaintext highlighter-rouge">""</code>), given that we anticipate
the <code class="language-plaintext highlighter-rouge">strand</code> to exist (signified by a non-empty string).</p>

<p>Before we go ahead and write our test case, let‚Äôs have a bit of a think around
the behaviour of our function in the case of an empty string. Should it:</p>

<ol>
  <li>Return a count of <code class="language-plaintext highlighter-rouge">0 0 0 0</code>, or</li>
  <li>Raise an exception?</li>
</ol>

<p>Usually, in situations like this, a decision like this already forms part of
our specification - but in our case, though, the exercise contains no indication
as to what is considered canonical, so we can choose either.</p>

<p>Let‚Äôs go with expecting a count of <code class="language-plaintext highlighter-rouge">0 0 0 0</code> for this one (there don‚Äôt appear to
be any significant benefits whichever one we choose).</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">it</span> <span class="s1">'returns a count of 0 0 0 0 for a strand with zero bases (empty string)'</span> <span class="k">do</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s1">''</span>
    <span class="n">nucleotide_count</span> <span class="o">=</span> <span class="s1">'0 0 0 0'</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>

    <span class="n">assert_equal</span> <span class="n">nucleotide_count</span><span class="p">,</span> <span class="n">result</span>
  <span class="k">end</span>
</code></pre></div></div>

<p>Let‚Äôs check our function‚Äôs behaviour:</p>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 33926

<span class="c"># Running:</span>

...

Finished <span class="k">in </span>0.000378s, 7936.5079 runs/s, 7936.5079 assertions/s.

3 runs, 3 assertions, 0 failures, 0 errors, 0 skips
</code></pre></div></div>

<p>Very nice.</p>

<hr />

<p>Are we done, now?</p>

<p>Not so fast.</p>

<p>There‚Äôs a requirement in our specification that we have ignored so far:</p>

<blockquote>
  <p>Given: A DNA string s of length <em>at most 1000 nt</em>.</p>
</blockquote>

<p>(Emphasis mine.)</p>

<p>Let‚Äôs quickly add a test case with an invalid length (&gt; 1000 nucleotides) to see
how our code behaves with against this requirement:</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">it</span> <span class="s1">'throws an exception if a strand is more than 1000nt long'</span> <span class="k">do</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s1">'A'</span> <span class="o">*</span> <span class="mi">1005</span>

    <span class="n">assert_raises</span><span class="p">(</span><span class="no">ArgumentError</span><span class="p">)</span> <span class="p">{</span> <span class="n">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span> <span class="p">}</span>
<span class="k">end</span>
</code></pre></div></div>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 15834

<span class="c"># Running:</span>

...F

Finished <span class="k">in </span>0.000579s, 6908.4628 runs/s, 6908.4628 assertions/s.

  1<span class="o">)</span> Failure:
nucleotide counting <span class="k">function</span><span class="c">#test_0004_throws an exception if a strand is more than 1000nt long [counting_nucleotides.rb:52]:</span>
ArgumentError expected but nothing was raised.

4 runs, 4 assertions, 1 failures, 0 errors, 0 skips
</code></pre></div></div>

<p>The test fails, as we were anticipating an exception but none was raised.</p>

<p>Let‚Äôs change our function to factor in this new requirement.</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>
    <span class="n">count_a</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_t</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_c</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_g</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">if</span> <span class="n">strand</span><span class="p">.</span><span class="nf">length</span> <span class="o">&gt;</span> <span class="mi">1000</span> <span class="k">then</span>
        <span class="k">raise</span> <span class="no">ArgumentError</span><span class="p">.</span><span class="nf">new</span><span class="p">(</span><span class="s1">'A strand of at most 1000nt is expected'</span><span class="p">)</span>
    <span class="k">end</span>

    <span class="n">strand</span><span class="p">.</span><span class="nf">split</span><span class="p">(</span><span class="s1">''</span><span class="p">).</span><span class="nf">each</span> <span class="k">do</span> <span class="o">|</span><span class="n">base</span><span class="o">|</span>
        <span class="k">if</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'A'</span> <span class="k">then</span>
            <span class="n">count_a</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elsif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'T'</span> <span class="k">then</span>
            <span class="n">count_t</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elsif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'C'</span> <span class="k">then</span>
            <span class="n">count_c</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elsif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'G'</span> <span class="k">then</span>
            <span class="n">count_g</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span>
            <span class="k">raise</span> <span class="no">ArgumentError</span><span class="p">.</span><span class="nf">new</span><span class="p">(</span><span class="s1">'Invalid character in strand '</span> <span class="o">+</span> <span class="n">base</span><span class="p">)</span>
        <span class="k">end</span>
    <span class="k">end</span>

    <span class="s2">"</span><span class="si">#{</span><span class="n">count_a</span><span class="si">}</span><span class="s2"> </span><span class="si">#{</span><span class="n">count_t</span><span class="si">}</span><span class="s2"> </span><span class="si">#{</span><span class="n">count_c</span><span class="si">}</span><span class="s2"> </span><span class="si">#{</span><span class="n">count_g</span><span class="si">}</span><span class="s2">"</span>
<span class="k">end</span>
</code></pre></div></div>

<p>Let‚Äôs see how our test does now:</p>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 14091

<span class="c"># Running:</span>

....

Finished <span class="k">in </span>0.000355s, 11267.6056 runs/s, 11267.6056 assertions/s.

4 runs, 4 assertions, 0 failures, 0 errors, 0 skips
</code></pre></div></div>

<p>Very nice. Everything passes now.</p>

<hr />

<p>Before we wrap this up, let‚Äôs make one final addition to our test case.</p>

<p>Do you remember how I mentioned that the exercise specification describes
a test case already?</p>

<p>We can incorporate this into our test cases. As a matter of fact, we can
substitute this one for the simpler positive case we had.</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">it</span> <span class="s1">'returns a count of 20 12 17 21 for the specification strand'</span> <span class="k">do</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s1">'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC'</span>
    <span class="n">nucleotide_count</span> <span class="o">=</span> <span class="s1">'20 12 17 21'</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>

    <span class="n">assert_equal</span> <span class="n">nucleotide_count</span><span class="p">,</span> <span class="n">result</span>
  <span class="k">end</span>
</code></pre></div></div>

<p>Let‚Äôs see how we fare against the story test case (the one given to us in
the specification).</p>

<div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>ruby counting_nucleotides.rb
Run options: <span class="nt">--seed</span> 17159

<span class="c"># Running:</span>

.....

Finished <span class="k">in </span>0.000428s, 11682.2430 runs/s, 11682.2430 assertions/s.

5 runs, 5 assertions, 0 failures, 0 errors, 0 skips
</code></pre></div></div>

<p>Looks like everything works as expected. üéâ</p>

<hr />

<p>Before I wrap up, I would be remiss if I did not mention that this particular
approach of designing code (<em>Test-Driven Development</em>) works very well when we
know the expected output of our code (say, when we know a lot about the domain,
or when our specification allows for examples that demonstrate expected input
and output).</p>

<p>It doesn‚Äôt work as great, however, when we don‚Äôt know what the output is (say,
for instance, when we do exploratory programming, as in the case of exploring
an API that‚Äôs given to us).</p>

<p>The complete code for this small exercise is listed below:</p>

<div class="language-rb highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">require</span> <span class="s1">'minitest/autorun'</span>

<span class="c1">##¬†IMPLEMENTATION CODE</span>

<span class="k">def</span> <span class="nf">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>
    <span class="n">count_a</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_t</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_c</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_g</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">if</span> <span class="n">strand</span><span class="p">.</span><span class="nf">length</span> <span class="o">&gt;</span> <span class="mi">1000</span> <span class="k">then</span>
        <span class="k">raise</span> <span class="no">ArgumentError</span><span class="p">.</span><span class="nf">new</span><span class="p">(</span><span class="s1">'A strand of at most 1000nt is expected'</span><span class="p">)</span>
    <span class="k">end</span>

    <span class="n">strand</span><span class="p">.</span><span class="nf">split</span><span class="p">(</span><span class="s1">''</span><span class="p">).</span><span class="nf">each</span> <span class="k">do</span> <span class="o">|</span><span class="n">base</span><span class="o">|</span>
        <span class="k">if</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'A'</span> <span class="k">then</span>
            <span class="n">count_a</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elsif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'T'</span> <span class="k">then</span>
            <span class="n">count_t</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elsif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'C'</span> <span class="k">then</span>
            <span class="n">count_c</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elsif</span> <span class="n">base</span> <span class="o">==</span> <span class="s1">'G'</span> <span class="k">then</span>
            <span class="n">count_g</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span>
            <span class="k">raise</span> <span class="no">ArgumentError</span><span class="p">.</span><span class="nf">new</span><span class="p">(</span><span class="s1">'Invalid character in strand '</span> <span class="o">+</span> <span class="n">base</span><span class="p">)</span>
        <span class="k">end</span>
    <span class="k">end</span>

    <span class="s2">"</span><span class="si">#{</span><span class="n">count_a</span><span class="si">}</span><span class="s2"> </span><span class="si">#{</span><span class="n">count_c</span><span class="si">}</span><span class="s2"> </span><span class="si">#{</span><span class="n">count_g</span><span class="si">}</span><span class="s2"> </span><span class="si">#{</span><span class="n">count_t</span><span class="si">}</span><span class="s2">"</span>
<span class="k">end</span>


<span class="c1">##¬†TEST CODE</span>

<span class="n">describe</span> <span class="s1">'nucleotide counting function'</span> <span class="k">do</span>
  <span class="c1"># Positive case</span>
  <span class="n">it</span> <span class="s1">'returns a count of 20 12 17 21 for the specification strand'</span> <span class="k">do</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s1">'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC'</span>
    <span class="n">nucleotide_count</span> <span class="o">=</span> <span class="s1">'20 12 17 21'</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>

    <span class="n">assert_equal</span> <span class="n">nucleotide_count</span><span class="p">,</span> <span class="n">result</span>
  <span class="k">end</span>

  <span class="c1"># Negative cases</span>
  <span class="n">it</span> <span class="s1">'throws an exception if a non-base encoding character is found in the strand'</span> <span class="k">do</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s1">'ATCGW'</span>
    
    <span class="n">assert_raises</span><span class="p">(</span><span class="no">ArgumentError</span><span class="p">)</span> <span class="p">{</span> <span class="n">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span> <span class="p">}</span>
  <span class="k">end</span>

  <span class="n">it</span> <span class="s1">'throws an exception if a strand is more than 1000nt long'</span> <span class="k">do</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s1">'A'</span> <span class="o">*</span> <span class="mi">1005</span>

    <span class="n">assert_raises</span><span class="p">(</span><span class="no">ArgumentError</span><span class="p">)</span> <span class="p">{</span> <span class="n">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span> <span class="p">}</span>
  <span class="k">end</span>

  <span class="c1"># Degenerate cases</span>
  <span class="n">it</span> <span class="s1">'returns a count of 0 0 0 0 for a strand with zero bases (empty string)'</span> <span class="k">do</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s1">''</span>
    <span class="n">nucleotide_count</span> <span class="o">=</span> <span class="s1">'0 0 0 0'</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">count_nucleotides</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>

    <span class="n">assert_equal</span> <span class="n">nucleotide_count</span><span class="p">,</span> <span class="n">result</span>
  <span class="k">end</span>
<span class="k">end</span>
</code></pre></div></div>

        </div>
        
      </li>
    
      
      

      <li>
        <header class="post-header">
          <h1 class="post-title">
            
              <a class="post-link" href="/2014/07/09/distro-forking-101/">Distro forking 101: How do you fork a Linux distro?</a>
            
          </h1>

          <p class="post-meta">
            Jul 9, 2014
            
              

 ‚Ä¢
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/linux/">linux</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/distro/">distro</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/fork/">fork</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/debian/">debian</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/arch/">arch</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/fedora/">fedora</a>
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  


            
            
          </p>
        </header>

        <div class="post-content">
          <h1>Defining the GNU/Linux distribution</h1>

<p>If you are here, we can safely assume that you already know what a <strong>GNU/Linux software
distribution</strong> is, but for completion‚Äôs sake, let‚Äôs just define so we all have the same context.</p>

<p>A GNU/Linux distribution is a collection of system and application software, packaged together
by the distribution‚Äôs developers, so that they are distributed in a nicely integrated bundle, ready
to be used by users and developers alike. Software typically included in such a distribution
ranges from a compiler toolchain, to the C library, to filesystem utilities to text editors.</p>

<p>As you can imagine, from the existence of several different GNU/Linux distributions, there are 
multiple ways that you could possibly combine all these different applications and their respective
configurations, not to mention that you could include even more specialised software, depending
on the target audience of the distribution (such as multimedia software for a distribution like 
<a href="http://ubuntustudio.org/">Ubuntu Studio</a> or penetration testing tools for a distribution such as
<a href="http://www.kali.org/">Kali Linux</a>)</p>

<h1>The ‚Äúf‚Äù word</h1>

<p>But even with such a great number of different software collections and their respective configurations
there still may not be one that appeals to your specific needs. That‚Äôs ok though, as you can still
customize each and every distribution to your specific liking. Extensive customization is known to
create a <em>differentiation point</em> known as a <strong>potential forking point</strong>.</p>

<p>Forking is a term that has been known to carry negative connotations. <a href="http://en.wikipedia.org/wiki/Fork_%28software_development%29">As wikipedia puts it</a>,</p>

<blockquote>
  <p>the term often implies not merely a development branch, but a split in the developer community
a form of schism.</p>
</blockquote>

<p>Historically, it has also been used as a leverage to coerce a project‚Äôs developers 
into merging code into their master branches
that they didn‚Äôt originally want to, or otherwise take a decision that they wouldn‚Äôt have taken
if not under the pressure of a <em>‚Äúfork‚Äù</em>. But why is it so?</p>

<p>You see, traditionally, forking a project meant a couple of things: For starters, there were now
two, identical projects, competing in the same solution space. 
Those two projects had different development hours and 
features or bug fixes going into them, and eventually, one of the two ended up being obsolete.
Apart from that forking also created an atmosphere of intense competition among the two projects.</p>

<p>However, in 2014, and the advent of the distributed version control systems such as <a href="http://git-scm.com/">git</a> 
and <a href="http://mercurial.selenic.com/">mercurial</a> and of the social coding websites such as <a href="http://www.github.com">Github</a> and <a href="http://www.bitbucket.org">Bitbucket</a>, the term is finally taking
on a more lax meaning, as just another code repository that may or may not enjoy major 
(or even minor, for that matter) development.</p>

<h2>Forking a GNU/Linux distribution</h2>

<p>So, up until now we have discussed what a GNU/Linux distribution is, and what a fork is. 
However, we haven‚Äôt discussed yet what it means to fork a GNU/Linux distribution.</p>

<p>You see, what differentiates each distro from the other ones, apart from the software collection 
that they contain, is the way in which they provide (and deploy) that software. Yes, we are talking about software packages and their respective package managers. Distributions from the Debian
(.deb) family are using <code class="language-plaintext highlighter-rouge">dpkg</code> along with <code class="language-plaintext highlighter-rouge">apt</code> or <code class="language-plaintext highlighter-rouge">synaptic</code> or <code class="language-plaintext highlighter-rouge">aptitude</code> or some other higher level
tool. RPM (.rpm) based distributions may use <code class="language-plaintext highlighter-rouge">rpm</code> with <code class="language-plaintext highlighter-rouge">yum</code> or <code class="language-plaintext highlighter-rouge">dnf</code> or  <code class="language-plaintext highlighter-rouge">zypper</code> or another higher level tool. 
Other distributions, not based on the aforementioned may choose to roll their own configuration
of packages and package managers, with Arch Linux using its own <code class="language-plaintext highlighter-rouge">pacman</code>, Sabayon uses its <code class="language-plaintext highlighter-rouge">entropy</code>
package manager, etc.</p>

<p>Now, naturally, if you want to customize an application to your liking, you have many ways in which
you could do that. One of them is downloading the tarball from the upstream‚Äôs website or ftp 
server, <code class="language-plaintext highlighter-rouge">./configure</code> it and then <code class="language-plaintext highlighter-rouge">make</code> and <code class="language-plaintext highlighter-rouge">make install</code> it. But if you do start customizing
lots of applications this way, it can become tedious and unwieldy too soon. After all, what did
that <code class="language-plaintext highlighter-rouge">make install</code> install exactly? Will the new update replace those files? What were your
configuration options? Did they replace the files the package manager installed?</p>

<p>In this case, it really pays off to learn <strong>packaging</strong> software for your distribution of choice.
What this means is to learn the format of packages your distribution‚Äôs package manager accepts
as well as how you could produce them. This way, instead of the <code class="language-plaintext highlighter-rouge">./configure &amp;&amp; make &amp;&amp; make install</code> 
cycle, you just have beautiful software packages, that you can control more tightly, you can 
update more easily and you can also distribute them to your friends if you so desire. As an added
bonus, now the package manager also knows about those files, and you can install, remove
or update them much more easily. What‚Äôs not to like?</p>

<p>After you have created some custom packages, you may also wish to create a repository
to contain them and update straight from that. <strong>Congratulations, you have created your custom
distribution, and a potential fork.</strong> While you are at it, if you really want to fork the distribution,
you could as well take the distribution‚Äôs <code class="language-plaintext highlighter-rouge">base</code> packages, customize them, rebuild them, and then
distribute them. <strong>Congratulations again, now you have your true GNU/Linux distribution fork</strong>.</p>

<h1>That seems easy. More specifically?</h1>

<p>Yes of course. Let‚Äôs take a look at how you might want to go about forking some well known
GNU/Linux distribution.</p>

<h2>Debian</h2>

<p>In Debian, your usual procedure if you wish to customize a package is the below:</p>

<ul>
  <li>First, you make sure you have the essential building software installed. <code class="language-plaintext highlighter-rouge">apt-get install build-essential devscripts debhelper</code></li>
  <li>Then you need to download the package‚Äôs build dependencies. <code class="language-plaintext highlighter-rouge">apt-get build-dep $package_name</code></li>
  <li>Now it‚Äôs time to download it‚Äôs sources, via <code class="language-plaintext highlighter-rouge">apt-get source $package_name</code></li>
  <li>Proceed with customizing it to your liking (update it, patch the sources, etc)</li>
  <li>Now it‚Äôs time to rebuild it. <code class="language-plaintext highlighter-rouge">debuild -us -uc</code></li>
</ul>

<p>Assuming all went fine, you should now have an <code class="language-plaintext highlighter-rouge">$package_name.deb</code> file in your current 
directory ready to be installed with <code class="language-plaintext highlighter-rouge">dpkg -i $package_name.deb</code>.</p>

<p>Please take note that the above is not an extensive treatise into debian packaging by any means.
If you want to build custom debian packages, here are some links:</p>

<ul>
  <li><a href="https://wiki.debian.org/IntroDebianPackaging">Debian wiki: intro to Debian packaging</a></li>
  <li><a href="http://people.connexer.com/~roberto/howtos/debcustomize">Roberto C Sanchez: Debian package customization how to</a></li>
  <li><a href="https://wiki.debian.org/HowToPackageForDebian">Debian wiki: How to package for Debian</a></li>
</ul>

<p>Now that you have your custom packages, it‚Äôs time to build a repository to contain them. There 
are many tools you can use to do that, including the official debian package archiving tool
known as <code class="language-plaintext highlighter-rouge">dak</code>, but if you want a personal repository without too much hassle, it‚Äôs better if you
use <code class="language-plaintext highlighter-rouge">reprepro</code>. I won‚Äôt go to full length on that here, <a href="http://www.debian-administration.org/article/286/Setting_up_your_own_APT_repository_with_upload_support">but instead I will lead you to a very
good guide to do so if you so desire</a></p>

<h2>Fedora</h2>

<p>Building packages for fedora is a procedure similar to the debian one. Fedora however is more
convenient in one aspect: <a href="http://download.fedoraproject.org/pub/fedora/linux/releases/20/Fedora/source/iso/Fedora-20-source-DVD.iso">It allows you to download a DVD image with all the sources in <code class="language-plaintext highlighter-rouge">.rpm</code> form
ready for you to customize and rebuild to your tastes.</a></p>

<p>Apart from that, the usual procedure is the following:</p>

<ul>
  <li>Download the <code class="language-plaintext highlighter-rouge">SRPM</code> (source RPM) via any means. You could do that using the <code class="language-plaintext highlighter-rouge">yumdownloader</code> utility, likewise <code class="language-plaintext highlighter-rouge">yumdownloader $package_name</code>. To use <code class="language-plaintext highlighter-rouge">yumdownloader</code>, you need
to have <code class="language-plaintext highlighter-rouge">yum-utils</code> installed.</li>
  <li>After you have downloaded the <code class="language-plaintext highlighter-rouge">SRPM</code>, next you have to unpack it: <code class="language-plaintext highlighter-rouge">rpm -i $package_name</code></li>
  <li>Next up, you customize the package to your liking (patch the sources, etc)</li>
  <li>Finally, you <code class="language-plaintext highlighter-rouge">cd</code> to the <code class="language-plaintext highlighter-rouge">SPECS</code> folder, and then <code class="language-plaintext highlighter-rouge">rpmbuild -ba $package.spec</code></li>
</ul>

<p>Again the above mentioned steps may not be 100% correct. If you want to go down this route,
see the following links for more information:</p>

<ul>
  <li><a href="http://wiki.centos.org/HowTos/RebuildSRPM">Centos wiki: Rebuild SRPM how-to</a></li>
  <li><a href="http://www.cyberciti.biz/faq/yum-download-source-packages-from-rhn/">cyberciti: yum Download all Source Packages from RedHat</a></li>
  <li><a href="http://bradthemad.org/tech/notes/patching_rpms.php">bradthemad.org: How to patch and rebuild an RPM package</a></li>
  <li><a href="http://www.rpm.org/max-rpm/ch-rpm-build.html">rpm.org: Chapter 11. Building packages</a></li>
  <li><a href="https://fedoraproject.org/wiki/How_to_create_an_RPM_package">Fedora wiki: How to create an RPM package</a></li>
</ul>

<p>Next up, is the repository creation step.</p>

<ul>
  <li>To create a yum repository, you need to <code class="language-plaintext highlighter-rouge">yum install createrepo</code>. 
After that you need to create a directory to use as the repository, likewise
<code class="language-plaintext highlighter-rouge">mkdir /var/ftp/repo/Fedora/19/{SRPMS, i386,x86_64)</code>.</li>
  <li>After that you move your i386 packages to <code class="language-plaintext highlighter-rouge">/var/ftp/repo/Fedora/19/i386</code>, and the rest
of the packages to their respective folders.</li>
  <li>Next step is adding a configuration file to <code class="language-plaintext highlighter-rouge">/etc/yum.repos.d/</code> that describes your repository
to yum.</li>
</ul>

<p>Again, not definitive, and for more information, take a look at these links:</p>

<ul>
  <li><a href="https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Deployment_Guide/sec-Yum_Repository.html">Redhat: Creating a yum repository</a></li>
  <li><a href="http://www.techrepublic.com/blog/linux-and-open-source/create-your-own-yum-repository">techrepublic: Creating your own yum repository</a></li>
  <li><a href="http://docs.fedoraproject.org/en-US/Fedora/14/html/Deployment_Guide/sec-Creating_a_Yum_Repository.html">Fedora documentation: Creating a yum repository</a></li>
</ul>

<h2>Arch Linux</h2>

<p><a href="https://www.archlinux.org/">Arch Linux</a>, at least in comparison to <code class="language-plaintext highlighter-rouge">.deb</code> and <code class="language-plaintext highlighter-rouge">.rpm</code> package
distribution families is very easy to customize to your liking. That‚Äôs to be expected though
as Arch Linux is a distribution that sells itself of the customization capabilities it offers to its user.</p>

<p>In essence, if you want to customize a package, the general process is this:</p>

<ul>
  <li>Download Arch tarball that contains the <code class="language-plaintext highlighter-rouge">PKGBUILD</code> file</li>
  <li>untar the tarball</li>
  <li>(Optional) download the upstream tarball referenced in the <code class="language-plaintext highlighter-rouge">PKGBUILD</code>, and modify it
 to your liking</li>
  <li>run <code class="language-plaintext highlighter-rouge">makepkg</code> in the folder containing the <code class="language-plaintext highlighter-rouge">PKGBUILD</code></li>
  <li>install (using pacman) the <code class="language-plaintext highlighter-rouge">.xz</code> file produced after the whole process is finished.</li>
</ul>

<p>In order to download the official {core | extra | community} packages, you need to run as root
<code class="language-plaintext highlighter-rouge">abs</code>. This will create a directory tree that contains the files required for building any package 
in the official repositories.</p>

<p>Next up, you can create a custom local repository with the <code class="language-plaintext highlighter-rouge">repo-add</code> tool, and then proceeding
with editing <code class="language-plaintext highlighter-rouge">/etc/pacman.conf</code> and adding an entry for your repository there. For more information:</p>

<ul>
  <li><a href="http://arch-stable.blogspot.gr/2012/02/make-your-own-local-repo-for-arch-linux.html">arch-stable: make your own local repo for arch linux</a></li>
  <li><a href="https://wiki.archlinux.org/index.php/makepkg">arch wiki: makepkg</a></li>
  <li><a href="https://wiki.archlinux.org/index.php/Pacman_tips">arch wiki: pacman tips</a></li>
  <li><a href="https://wiki.archlinux.org/index.php/Arch_Build_System">arch wiki: Arch Build System</a></li>
  <li><a href="https://wiki.archlinux.org/index.php/Pacman#Repositories">arch wiki: Pacman#Repositories</a></li>
</ul>

<h1>To fork or not to fork?</h1>

<p>Well, that‚Äôs not an easy question to answer. My opinion is that it‚Äôs extremely educational to
do a <strong><em>soft</em></strong> fork, clone the distribution‚Äôs core repository, and for some time maintain your own
distribution based on it, that is, update and customize all the repositories. Do that for some months,
then go back to using your distribution of choice now that you are enlightened with how it works
under the hood. The reason this is very educational is that it will teach you the ins and outs of
your distribution, teach you about <strong>all</strong> the software in it, how it integrates, what its role is.
It will teach you packaging which is a tremendously undervalued skill, as you can customize
your experience to your liking, and it will make you appreciate the effort going into maintaining
the distribution.</p>

<p>As for doing a <strong>hard</strong> fork, that is creating your own distribution, that you commit to maintaining
it for a long time, my opinion is that it‚Äôs simply not worth it. Maintaining a distribution, be it
by yourself, or with your friends, is a tremendous amount of work, that‚Äôs not worth it unless
you have other goals you want to achieve by that. If all you want to do is to customize your 
distribution of choice to your liking, then go ahead, learn packaging for it, customize-package
the applications you want, then create your own repo - but always track the upstream. Diverging
too much from the upstream is not worth the hassle, as you will end up spending more time
maintaining than using the distribution in the end.</p>

<h2>tl;dr:</h2>

<p>If you want to do a small scale, private fork in order to see what‚Äôs under the hood of your Linux
distro; by all means go ahead.</p>

<p>If you want to do a large scale, public fork, then take your time to calculate the effort, if it‚Äôs worth it,
and if you could just help the upstream distribution implement the features you want.</p>

        </div>
        
      </li>
    
      
      

      <li>
        <header class="post-header">
          <h1 class="post-title">
            
              <a class="post-link" href="/2014/06/20/how-the-compiler/">How the Compiler, the Library and the Kernel work - Part 3</a>
            
          </h1>

          <p class="post-meta">
            Jun 20, 2014
            
              

 ‚Ä¢
  
    
    
      
    
      
    
      
    
      
        <a href="/tags/compiler/">compiler</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/kernel/">kernel</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/libc/">libc</a>
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  


            
            
          </p>
        </header>

        <div class="post-content">
          <p>In the last part of this series, we talked about the compiler‚Äôs composition, including the assembler
and the linker. We showed what happens when the compiler runs, and what‚Äôs the output
of translation software such as <code class="language-plaintext highlighter-rouge">cc1</code> or <code class="language-plaintext highlighter-rouge">as</code> etc. In this final part of the series, we are going
to talk about the C library, how our programs interface with it, and how it interfaces with 
the kernel.</p>

<h1>The C Standard Library</h1>

<p>The C Standard Library is pretty much a part of every UNIX like operating system. It‚Äôs basically
a collection of code, including functions, macros, type definitions etc, in order to provide facilities
such as string handling (<code class="language-plaintext highlighter-rouge">string.h</code>), mathematical computations (<code class="language-plaintext highlighter-rouge">math.h</code>), input and output
(<code class="language-plaintext highlighter-rouge">stdio.h</code>), etc.</p>

<p>GNU/Linux operating systems are generally using the <a href="http://www.gnu.org/software/libc/libc.html">GNU C Library implementation(GLIBC)</a>,
but it‚Äôs common to find other C libraries being used (especially in embedded systems) such as 
<a href="http://www.uclibc.org">uClibC</a>, <a href="http://sources.redhat.com/newlib">newlib</a>, or in the case
of Android/Linux systems <a href="https://android.googlesource.com/platform/bionic.git">Bionic</a>.
BSD style operating systems usually have their own implementation of a C library.</p>

<h2>So, how does one ‚Äúuse‚Äù the C Standard Library?</h2>

<p>So, now that we are acquainted with the C Library, how do you make use of it, you ask? The answer is:
<strong>automagically</strong> :). Hold on right there; that‚Äôs not exactly a hyperbole. You see, when you
write a basic C program, you usually <code class="language-plaintext highlighter-rouge">#include &lt;some_header.h&gt;</code> and then continue with
using the code declared in that header. We have explained in the previous part of this series
that when we use a function, say <code class="language-plaintext highlighter-rouge">printf()</code>, in reality it‚Äôs the linker that does the hard work 
and allows us to use this function, by linking our program against the <code class="language-plaintext highlighter-rouge">libc</code>‚Äôs <code class="language-plaintext highlighter-rouge">so</code> (shared 
object). So in essence, when you need to use the C Standard Library, you just <code class="language-plaintext highlighter-rouge">#include</code>
headers that belong to it, and the linker will resolve the references to the code included.</p>

<p>Apart from the functions that are defined in the Standards however, a C Library might also
implement further functionality. For example, the Standards don‚Äôt say anything about networking.
As a matter of fact, most libraries today may implement not only what‚Äôs in the C Standards,
but may also choose to comply with the requirements of the POSIX C library, which is a superset
of the C Standard library.</p>

<h2>Ok, and how does the C Library manage to provide these services?</h2>

<p>The answer to this question is simple: Some of the services that the library provides, it does so
without needing any sort of special privileges, being normal, userspace C code, while others
need to ask the Operating‚Äôs system Kernel to provide these facilities for the library.</p>

<p>How does it do so? By calling some functions exported by the kernel to provide certain functionality
 named <strong>system calls</strong>. System calls are the fundamental interface between a userspace
application and the Operating System Kernel. For example consider this:</p>

<p>You might have a program that has code like this at one point: <code class="language-plaintext highlighter-rouge">fd = open("log.txt", "w+");</code>. That
<code class="language-plaintext highlighter-rouge">open</code> function is provided by the C Library, but the C Library itself can not execute all of the
functionality that‚Äôs required to open a file, so it may call a <code class="language-plaintext highlighter-rouge">sys_open()</code> system call that will
ask the kernel to do what‚Äôs required to load the file. In this case we say that the library‚Äôs <code class="language-plaintext highlighter-rouge">open</code>
call acts as a wrapper function of the system call.</p>

<h1>Epilogue</h1>

<p>In this final part of our series, we saw how our applications interface with the C Standard Library
available in our system, and how the Library itself interfaces with the Operating system kernel
to provide the required services needed by the userspace applications.</p>

<h2>Further Reading:</h2>

<p>If you want to take a look at the System Call interface in the Linux Operating System, you could
always see the <a href="http://man7.org/linux/man-pages/man2/syscalls.2.html">man page for the Linux system calls</a></p>

        </div>
        
      </li>
    
      
      

      <li>
        <header class="post-header">
          <h1 class="post-title">
            
              <a class="post-link" href="/2014/04/28/introduction-to-xv6-adding-a-new-system-call/">Introduction to xv6: Adding a new system call.</a>
            
          </h1>

          <p class="post-meta">
            Apr 28, 2014
            
              

 ‚Ä¢
  
    
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/kernel/">kernel</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/ensidia/">ensidia</a>,
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  
    
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
        <a href="/tags/xv6/">xv6</a>
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
      
    
  


            
            
          </p>
        </header>

        <div class="post-content">
          <h1>xv6: An introduction</h1>

<p>If you are like me, a low level pc programmer, it‚Äôs hard not to have heard
of <strong>xv6</strong>. <a href="http://pdos.csail.mit.edu/6.828/2012/xv6.html">xv6</a>, for those
who haven‚Äôt really heard of it, is a <em>UNIX version 6</em> clone, designed
at MIT to help teach operating systems.</p>

<p>The reasoning behind doing this was fairly simple: <a href="http://www.lemis.com/grog/Documentation/Lions/">Up until that point, MIT
had used John Lions‚Äô famous commentary on the Sixth Edition of UNIX</a>. But V6 was challenging due to a 
number of reasons. To begin with, it was written in a near ancient version
of C (pre K&amp;R), and apart from that, it contained PDP-11 assembly
(a legendary machine for us UNIX lovers, but ancient nonetheless), which
didn‚Äôt really help the students that had to study both PDP-11 and the
(more common) x86 architecture to develop another (exokernel) operating
system on.</p>

<p>So, to make things much more simpler, professors there decided to 
roll with a clone of UNIX version 6, that was x86 specific, 
written in ANSI C and supported multiprocessor machines.</p>

<p>For a student (or a programmer interested in operating systems), xv6 is 
a unique opportunity to introduce himself to kernel hacking and to the
architecture of UNIX like systems. At about 15k lines of code (iirc), 
including the (primitive) libraries, the userland and the kernel,
it‚Äôs very easy (or well, at least easier than production scale UNIX like
systems) to grok, and it‚Äôs also very easy to expand on. It also helps
tremendously that xv6 as a whole has magnificent documentation, not only
from MIT, but from other universities that have adopted xv6 for use in their
operating systems syllabus.</p>

<h2>An introduction to Ensidia: my very personal xv6 fork</h2>

<p>When I first discovered xv6 I was ecstatic. For the reasons mentioned above
I couldn‚Äôt lose on the opportunity to fork xv6 and use it as a personal
testbed for anything I could feel like exploring or testing out.</p>

<p>As a matter of fact, when I first discovered xv6, <a href="https://github.com/NlightNFotis/Fotix">I had just finished 
implementing (the base of) my own UNIX like operating system, named fotix</a>, 
and the timing of my discovery was great. xv6 had done what I had done,
and also implemented most of what I was planning to work on fotix
(for example, elf file loading), and it was a solid base for further 
development. It also had a userland, which fotix at the time didn‚Äôt have.</p>

<p>After I forked xv6, I spent some time familiriazing myself with the code.
I also cleaned up the source code quite a bit, structuring the code in a 
BSD like folder structure, instead of having all of the code in the same 
folder and made various small scale changes.</p>

<p>After that for quite some time, I had left ensidia alone and didn‚Äôt touch
it much. However, I always felt like I wanted to develop it a bit more
and get to play with its code in interesting ways. I was trying to think of
a great way to get started with kernel hacking on it, in a simple way, to 
get more acquainted with the kernel, and found an interesting pdf with 
interesting project ideas for it. One of them was to add a system call.
I figured out that would be an interesting and quick hack, so hey, why not?</p>

<h1>Getting started with kernel hacking on xv6: Adding the system call.</h1>

<p>The system call I decided to introduce was the suggested one. It was
fairly simple sounding too. You have to <strong>introduce a new system call
that returns the number of total system calls that have taken place
so far</strong>. So let‚Äôs see how I went about implementing it:</p>

<h2>An introduction to system calls in xv6</h2>

<p>First of all, we should provide some context about what system calls are,
how they are used, and how they are implemented in xv6.</p>

<p>A system call is a function that a userspace application will use, so as
to ask for a specific service to be provided by the operating system. For
instance with an <code class="language-plaintext highlighter-rouge">sbrk(n)</code> system call, a process can ask the kernel to
grow its heap space by n bytes. Another example is the well known <code class="language-plaintext highlighter-rouge">fork()</code>
system call in the UNIX world, that‚Äôs used to create a new process by 
cloning the caller process.</p>

<p>The way applications signal the kernel that they need that service is
by issueing a software <em>interrupt</em>. An <em>interrupt</em> is a signal generated
that notifies the processor that it needs to stop what its currently doing,
and handle the interrupt. This mechanism is also used to notify the processor
that information it was seeking from the disks is in some buffer, ready to
be extracted and processed, or, that a key was pressed in the keyboard. This
is called a hardware interrupt.</p>

<p>Before the processor stops to handle the interrupt generated, it needs to 
save the current state, so that it can resume the execution in this context
after the interrupt has been handled.</p>

<p>The code that calls a system call in xv6 looks like this:</p>

<figure class="highlight"><pre><code class="language-gas" data-lang="gas"># exec(init, argv)
 .globl start
 start:
   pushl $argv
   pushl $init
   pushl $0  // where caller pc would be
   movl $SYS_exec, %eax
   int $T_SYSCALL</code></pre></figure>

<p>In essence, it pushes the argument of the call to the stack, and puts
the system call number (in the above code, that‚Äôs <code class="language-plaintext highlighter-rouge">$SYS_exec</code>) into <code class="language-plaintext highlighter-rouge">%eax</code>.
The number is used to match the entry in an array that holds pointers to
all the system calls. After that, it generates a software interrupt, with
a code (in this case <code class="language-plaintext highlighter-rouge">$T_SYSCALL</code>) that‚Äôs used to index the interrupt
descriptor tables and find the appropriate interrupt handler.</p>

<p>The code that is specific to find the appropriate interrupt handler is
called <code class="language-plaintext highlighter-rouge">trap()</code> and is available in the file <code class="language-plaintext highlighter-rouge">trap.c</code>. If <code class="language-plaintext highlighter-rouge">trap()</code> check‚Äôs
out the trapnumber in the generated trapframe (a structure that represents
the processor‚Äôs state at the time that the trap happened) to be equal to
<code class="language-plaintext highlighter-rouge">T_SYSCALL</code>, it calls <code class="language-plaintext highlighter-rouge">syscall()</code> (the software interrupt handler)
 that‚Äôs available in <code class="language-plaintext highlighter-rouge">syscall.c</code></p>

<figure class="highlight"><pre><code class="language-c" data-lang="c"><span class="c1">// This is the part of trap that</span>
<span class="c1">// calls syscall()</span>
<span class="kt">void</span>
<span class="nf">trap</span><span class="p">(</span><span class="k">struct</span> <span class="n">trapframe</span> <span class="o">*</span><span class="n">tf</span><span class="p">)</span>
<span class="p">{</span>
  <span class="k">if</span><span class="p">(</span><span class="n">tf</span><span class="o">-&gt;</span><span class="n">trapno</span> <span class="o">==</span> <span class="n">T_SYSCALL</span><span class="p">){</span>
    <span class="k">if</span><span class="p">(</span><span class="n">proc</span><span class="o">-&gt;</span><span class="n">killed</span><span class="p">)</span>
      <span class="n">exit</span><span class="p">();</span>
    <span class="n">proc</span><span class="o">-&gt;</span><span class="n">tf</span> <span class="o">=</span> <span class="n">tf</span><span class="p">;</span>
    <span class="n">syscall</span><span class="p">();</span>
    <span class="k">if</span><span class="p">(</span><span class="n">proc</span><span class="o">-&gt;</span><span class="n">killed</span><span class="p">)</span>
      <span class="n">exit</span><span class="p">();</span>
    <span class="k">return</span><span class="p">;</span>
  <span class="p">}</span></code></pre></figure>

<p><code class="language-plaintext highlighter-rouge">syscall()</code> is finally the function that checks out <code class="language-plaintext highlighter-rouge">%eax</code> to get the 
number of the system call (to index the array with the system 
call pointers), and execute the code corresponding to that system call.</p>

<p>The implementation of system calls in xv6 is under two files. The first one
is <code class="language-plaintext highlighter-rouge">sysproc.c</code>, and is the one containing the implementation of system calls
correspondent to processes, and <code class="language-plaintext highlighter-rouge">sysfile.c</code> that contains the implementation
of system calls regarding the file system.</p>

<h2>The specific implementation of the <code class="language-plaintext highlighter-rouge">numcalls()</code> system call</h2>

<p>To implement the system call itself is simple. I did so with a global variable
in <code class="language-plaintext highlighter-rouge">syscall.c</code> called <code class="language-plaintext highlighter-rouge">syscallnum</code>, that‚Äôs incremented everytime 
<code class="language-plaintext highlighter-rouge">syscall()</code>, calls a system call function, that is, the system call
is valid.</p>

<p>Next we just need a function, the system call implementation that returns
that number to the userspace program that asks for it. Below is the 
function itself, and <code class="language-plaintext highlighter-rouge">syscall()</code> after our change.</p>

<figure class="highlight"><pre><code class="language-c" data-lang="c"><span class="c1">// return the number of system calls that have taken place in</span>
<span class="c1">// the system</span>
<span class="kt">int</span>
<span class="nf">sys_numcalls</span><span class="p">(</span><span class="kt">void</span><span class="p">)</span>
<span class="p">{</span>
    <span class="k">return</span> <span class="n">syscallnum</span><span class="p">;</span>
<span class="p">}</span></code></pre></figure>

<figure class="highlight"><pre><code class="language-c" data-lang="c"><span class="c1">// The syscall() implementation after</span>
<span class="c1">// our change</span>
<span class="kt">void</span>
<span class="nf">syscall</span><span class="p">(</span><span class="kt">void</span><span class="p">)</span>
<span class="p">{</span>
  <span class="kt">int</span> <span class="n">num</span><span class="p">;</span>

  <span class="n">num</span> <span class="o">=</span> <span class="n">proc</span><span class="o">-&gt;</span><span class="n">tf</span><span class="o">-&gt;</span><span class="n">eax</span><span class="p">;</span>
  <span class="k">if</span><span class="p">(</span><span class="n">num</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="o">&amp;&amp;</span> <span class="n">num</span> <span class="o">&lt;</span> <span class="n">NELEM</span><span class="p">(</span><span class="n">syscalls</span><span class="p">)</span> <span class="o">&amp;&amp;</span> <span class="n">syscalls</span><span class="p">[</span><span class="n">num</span><span class="p">])</span> <span class="p">{</span>
    <span class="n">syscallnum</span><span class="o">++</span><span class="p">;</span> <span class="c1">// increment the syscall counter</span>
    <span class="n">proc</span><span class="o">-&gt;</span><span class="n">tf</span><span class="o">-&gt;</span><span class="n">eax</span> <span class="o">=</span> <span class="n">syscalls</span><span class="p">[</span><span class="n">num</span><span class="p">]();</span>
  <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
    <span class="n">cprintf</span><span class="p">(</span><span class="s">"%d %s: unknown sys call %d</span><span class="se">\n</span><span class="s">"</span><span class="p">,</span>
            <span class="n">proc</span><span class="o">-&gt;</span><span class="n">pid</span><span class="p">,</span> <span class="n">proc</span><span class="o">-&gt;</span><span class="n">name</span><span class="p">,</span> <span class="n">num</span><span class="p">);</span>
    <span class="n">proc</span><span class="o">-&gt;</span><span class="n">tf</span><span class="o">-&gt;</span><span class="n">eax</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">;</span>
  <span class="p">}</span>
<span class="p">}</span></code></pre></figure>

<p>After that was done, the next few things that were needed to be done
were fairly straight forward. We had to add an index number for the new
system call in <code class="language-plaintext highlighter-rouge">syscall.h</code>, expose it to user proccesses via <code class="language-plaintext highlighter-rouge">user.h</code>,
and add a new macro to <code class="language-plaintext highlighter-rouge">usys.S</code> that defines an asm routine that calls
that specific system call, and change the makefile to facilitate our change
. After doing so we had to write a userspace testing program to test our changes.</p>

<p>The result after doing all this is below :)</p>

<figure class="highlight"><pre><code class="language-sh" data-lang="sh">cpu1: starting
cpu0: starting
init: starting sh
<span class="nv">$ </span><span class="nb">ls</span>
<span class="nb">.</span>              1 1 512
..             1 1 512
README         2 2 2209
<span class="nb">cat            </span>2 3 9725
<span class="nb">echo           </span>2 4 9254
forktest       2 5 5986
<span class="nb">grep           </span>2 6 10873
init           2 7 9579
<span class="nb">kill           </span>2 8 9246
<span class="nb">ln             </span>2 9 9240
<span class="nb">ls             </span>2 10 10832
<span class="nb">mkdir          </span>2 11 9315
<span class="nb">rm             </span>2 12 9308
sh             2 13 16600
stressfs       2 14 9790
usertests      2 15 37633
<span class="nb">wc             </span>2 16 10207
zombie         2 17 9028
syscallnum     2 18 9144
console        3 19 0
<span class="nv">$ </span>syscallnum
The total number of syscalls so far is 643
<span class="nv">$ </span>syscallnum
The total number of syscalls so far is 705
<span class="nv">$ </span>syscallnum
The total number of syscalls so far is 767
<span class="nv">$ </span>syscallnum
The total number of syscalls so far is 829</code></pre></figure>

<h1>Epilogue</h1>

<p>I usually end my blog posts with an epilogue. Although this is a post
that doesn‚Äôt necesarilly need one, <strong>I wanted to write one just to say to you
that you should try kernel hacking</strong>, <em>that is programming jargon for
programming an operating system kernel</em>, because it‚Äôs an experience that
undoubtedly will teach you a great deal of things about how your computer
actually works.</p>

<p>Last but not least, take a look at the ongoing work on <a href="https://github.com/NlightNFotis/Ensidia">Ensidia, my fork
of xv6</a>. To see this particular
work, take a look at the <code class="language-plaintext highlighter-rouge">syscall</code> branch.</p>

<h2>References</h2>
<ul>
  <li><a href="http://zoo.cs.yale.edu/classes/cs422/2010/reference">CS422/522: Operating systems, Yale</a></li>
  <li><a href="http://zoo.cs.yale.edu/classes/cs422/2010/xv6-book/fscall.pdf">Chapter 8, File System calls, xv6 reference, Yale</a></li>
  <li><a href="http://zoo.cs.yale.edu/classes/cs422/2010/xv6-book/trap.pdf">Chapter 3, System calls, exceptions and interrupts, Yale</a></li>
  <li><a href="http://pdos.csail.mit.edu/6.828/2012/xv6/book-rev7.pdf">xv6 Documentation, MIT csail</a></li>
</ul>

        </div>
        
      </li>
    
  </ul>

  
  <div class="pagination">
    
      <a class="previous" href="/posts/3/">&laquo; Older</a>
    

    
      <a class="next" href="/">Newer &raquo;</a>
    
  </div>



</div>

      </div>
    </main>

    <footer class="site-footer">

  <div class="wrapper">

    <p>
      

&copy; Fotis Koutoulakis - Powered by <a href="https://jekyllrb.com">Jekyll</a> &amp; <a href="https://github.com/yous/whiteglass">whiteglass</a> - Subscribe via <a href="https://nlightnfotis.github.io/feed.xml">RSS</a>

    </p>

  </div>

</footer>


  </body>

</html>
